File: backend/src/main/kotlin/com/mongle/common/exception/ErrorCode.kt
package com.mongle.common.exception

import com.mongle.common.Messages
import org.springframework.http.HttpStatus

/**
 * 애플리케이션 에러 코드. 새 에러 유형은 여기에 추가한다.
 *
 * message 는 사용자에게 그대로 보여줄 수 있는 §12.5 문구의 기본값이다.
 * 값이 들어가는 문구(글자수·칩 개수 등)는 throw 시점에 Messages 함수로 만들어 override 한다.
 */
enum class ErrorCode(
    val status: HttpStatus,
    val message: String,
) {
    INVALID_INPUT(HttpStatus.BAD_REQUEST, "잘못된 입력입니다."),
    REQUIRED_FIELD(HttpStatus.BAD_REQUEST, Messages.REQUIRED_NAME),
    LENGTH_EXCEEDED(HttpStatus.BAD_REQUEST, "글자수를 초과했어요."),
    DUPLICATE(HttpStatus.CONFLICT, Messages.DUPLICATE),
    FUTURE_DATE(HttpStatus.BAD_REQUEST, Messages.FUTURE_DATE),
    DATE_ORDER(HttpStatus.BAD_REQUEST, Messages.DATE_ORDER),
    SELECTION_LIMIT(HttpStatus.BAD_REQUEST, Messages.SELECTION_LIMIT),
    CHIP_LIMIT(HttpStatus.BAD_REQUEST, "칩 개수 상한을 넘었어요."),
    CATEGORY_REQUIRED(HttpStatus.BAD_REQUEST, Messages.CATEGORY_REQUIRED),
    UNAUTHORIZED(HttpStatus.UNAUTHORIZED, "로그인이 필요해요."),
    NOT_FOUND(HttpStatus.NOT_FOUND, "리소스를 찾을 수 없습니다."),
    UNSUPPORTED_IMAGE_TYPE(HttpStatus.BAD_REQUEST, Messages.UNSUPPORTED_IMAGE_TYPE),
    IMAGE_TOO_LARGE(HttpStatus.BAD_REQUEST, Messages.IMAGE_TOO_LARGE),
    SAVE_FAILED(HttpStatus.INTERNAL_SERVER_ERROR, Messages.SAVE_FAILED),
    INTERNAL_ERROR(HttpStatus.INTERNAL_SERVER_ERROR, "서버 오류가 발생했습니다."),
}

File: backend/src/common/exception/error-code.ts
import { HttpStatus } from '@nestjs/common';

export enum ErrorCode {
  INVALID_INPUT = 'INVALID_INPUT',
  REQUIRED_FIELD = 'REQUIRED_FIELD',
  LENGTH_EXCEEDED = 'LENGTH_EXCEEDED',
  DUPLICATE = 'DUPLICATE',
  FUTURE_DATE = 'FUTURE_DATE',
  DATE_ORDER = 'DATE_ORDER',
  SELECTION_LIMIT = 'SELECTION_LIMIT',
  CHIP_LIMIT = 'CHIP_LIMIT',
  CATEGORY_REQUIRED = 'CATEGORY_REQUIRED',
  UNAUTHORIZED = 'UNAUTHORIZED',
  NOT_FOUND = 'NOT_FOUND',
  INTERNAL_ERROR = 'INTERNAL_ERROR',
}

export interface ErrorDefinition {
  status: HttpStatus;
  message: string;
}

export const ERROR_DEFINITIONS: Record<ErrorCode, ErrorDefinition> = {
  [ErrorCode.INVALID_INPUT]: { status: HttpStatus.BAD_REQUEST, message: '잘못된 입력입니다.' },
  [ErrorCode.REQUIRED_FIELD]: { status: HttpStatus.BAD_REQUEST, message: '이름을 입력해 주세요.' },
  [ErrorCode.LENGTH_EXCEEDED]: { status: HttpStatus.BAD_REQUEST, message: '글자수를 초과했어요.' },
  [ErrorCode.DUPLICATE]: { status: HttpStatus.CONFLICT, message: '이미 있는 항목이에요.' },
  [ErrorCode.FUTURE_DATE]: { status: HttpStatus.BAD_REQUEST, message: '오늘보다 미래일 수는 없어요.' },
  [ErrorCode.DATE_ORDER]: {
    status: HttpStatus.BAD_REQUEST,
    message: '마지막 만난 날은 처음 만난 날 이후여야 해요.',
  },
  [ErrorCode.SELECTION_LIMIT]: {
    status: HttpStatus.BAD_REQUEST,
    message: '선택할 수 있는 최대 개수를 넘었어요.',
  },
  [ErrorCode.CHIP_LIMIT]: { status: HttpStatus.BAD_REQUEST, message: '칩 개수 상한을 넘었어요.' },
  [ErrorCode.CATEGORY_REQUIRED]: {
    status: HttpStatus.BAD_REQUEST,
    message: '카테고리는 최소 1개가 필요해요.',
  },
  [ErrorCode.UNAUTHORIZED]: { status: HttpStatus.UNAUTHORIZED, message: '로그인이 필요해요.' },
  [ErrorCode.NOT_FOUND]: { status: HttpStatus.NOT_FOUND, message: '리소스를 찾을 수 없습니다.' },
  [ErrorCode.INTERNAL_ERROR]: {
    status: HttpStatus.INTERNAL_SERVER_ERROR,
    message: '서버 오류가 발생했습니다.',
  },
};
